### ISSUE TITLE
aarch64: error: inline assembly requires more registers than available

### ISSUE THREAD

**Issue by am11** [libunwind/libunwind#702]
Hm, clang has a different error:
```
error: inline assembly requires more registers than available
   65 |         "ldr x0,  %[x0]\n\t"
      |         ^
```

_Originally posted by @sfc-gh-sgiesecke in https://github.com/libunwind/libunwind/issues/608#issuecomment-1708458429_

---

Hitting the same issue in dotnet-runtime, where arm64 legs are cross-compiled using clang.

**Comment by am11** [comment]
Looks like gcc also fails with -O0 here:

from https://github.com/libunwind/libunwind/actions/runs/7543438232/job/20534462217
```
aarch64/Gos-linux.c:41:7: error: 'asm' operand has impossible constraints
   41 |       __asm__ __volatile__ (
      |       ^~~~~~~
```

clang fails with all -O{0,2,3}.. it can only handle 30 registers in input operands at a time. Keeping any 30 in https://github.com/libunwind/libunwind/blob/b603e634d27aabf5797d95055003c9947f5195ab/src/aarch64/Gos-linux.c#L70-L94 (and adjusting template and clobbers above and below) fixes the error.

@bregma, this is a build regression in v1.8.0. Given the limitations of both clang and gcc,  it seems like we have three options:
1. split this block in two parts.
2. switch from inline assembly to dedicated assembly file .S
3. revert the original change and rework it for the next release

**Comment by TannerFirl** [comment]
I'm hitting this error, so won't be able to upgrade to v1.8 til its resolved

**Comment by am11** [comment]
I reverted https://github.com/libunwind/libunwind/pull/503 in dotnet/runtime to fix it like this: [dotnet/runtime@`d63c235` (#96969)](https://github.com/dotnet/runtime/pull/96969/commits/d63c235756d4c46b061bd503a2c47207df6b3324).

**Comment by jamel** [comment]
Hit similar issue when compile Libunwind with Clang-14.0.6 on Arch64 (graviton) Amazon Linux 2.

Here is the exact compilation command that fails:
```
$ clang -DHAVE_CONFIG_H -I. -I../include -I../include -I../include/tdep-aarch64 -I. -D_GNU_SOURCE -DNDEBUG -Wall -Wsign-compare -D__EXTENSIONS__ -c aarch64/Los-linux.c -o aarch64/.libs/Los-linux.o
In file included from aarch64/Los-linux.c:4:
aarch64/Gos-linux.c:45:9: error: inline assembly requires more registers than available
        "ldr x0,  %[x0]\n\t"
        ^
aarch64/Gos-linux.c:45:9: error: inline assembly requires more registers than available
aarch64/Gos-linux.c:45:9: error: inline assembly requires more registers than available
aarch64/Gos-linux.c:45:9: error: inline assembly requires more registers than available
aarch64/Gos-linux.c:45:9: error: inline assembly requires more registers than available
aarch64/Gos-linux.c:45:9: error: inline assembly requires more registers than available
aarch64/Gos-linux.c:45:9: error: inline assembly requires more registers than available
aarch64/Gos-linux.c:45:9: error: inline assembly requires more registers than available
aarch64/Gos-linux.c:45:9: error: inline assembly requires more registers than available
aarch64/Gos-linux.c:45:9: error: inline assembly requires more registers than available
10 errors generated.
```

**Comment by sdebnath** [comment]
This issue is tagged for 1.9.0. Would be great to get a fix as a patch in 1.8 series. Same issue as #742
